### DISCUSSION TITLE
Robust handling of version and entity ID overflow

### DISCUSSION THREAD

**Discussion by emsai** [skypjack/entt#875] [Ideas]
I am interested in an entity implementation that handles version overflow and maxing out entity ids more robustly.

As noted in https://github.com/skypjack/entt/discussions/761, long running applications that continuously delete and create entities can run into issues with the version number associated with an entity overflowing and incorrectly causing destroyed entities to be reported as valid. The behavior when the entity ID overflows is also undefined. 

Here is a potential way of eliminating the version overflow issue. For simplicity, I have described it using shared pointers for the entities, but this could be adapted to an implementation that stores the underlying data in a compact representation, using similar techniques to what is done for the registry vector. The code is just meant to give an idea of the approach and isn't intended as an actual implementation.

This implementation matches what entt does already, except that it separates the storage of the entity value from the allocation in the registry's vector, which eliminates the need for versions and issues associated with versions overflowing.

```
// Assume the registry has the following member variables:
std::vector<std::shared_ptr<uint32_t>> entities;
uint32_t head = std::numeric_limits<uint32_t>::max();

// Entity creation
std::shared_ptr<const uint32_t> create() {
      std::shared_ptr<uint32_t> entity;
      if (head == std::numeric_limits<uint32_t>::max()) {
               entity = std::make_shared<uint32_t>(entities.size());
               entities.push_back(entity);
      } else {
               entity = entities[head];
               uint32_t temp = head;
               head = *entities[head];
               *entities[head] = temp;
      }
      return entity;
}

// Entity destruction
void destroy() {
      uint32_t new_entity_index = *entity;
      entities[new_entity_index] = std::make_shared<uint32_t>(head);
      head = new_entity_index;
      *entity = std::numeric_limits<uint32_t>::max();
}

void is_valid(std::shared_ptr<const uint32_t> entity) {
    return *entity != std::numeric_limits<uint32_t>::max();
}
```

With this approach, version numbers are not needed anymore. Just one value (already reserved as null in entt) needs to be reserved to indicate that an entity is invalid. So more bits are freed up for the ID, which could help eliminate reaching the simultaneous entity ID limit as well.

Any thoughts on this approach and how feasible it would be to include in entt? Would it be possible to abstract the entity type used by the registry to accommodate this approach in addition to the existing behavior?

I would be interested in hearing about other suggestions for improving robustness as well.

**Comment by skypjack** [comment]
> The behavior when the entity ID overflows is also undefined.

What do you mean exactly? Currently, it just wraps to 0. It's pretty well defined.

> With this approach, version numbers are not needed anymore. Just one value (already reserved as null in entt) needs to be reserved to indicate that an entity is invalid.

There is a difference between an invalid entitiy and an entity with an updated version though. That is more or less the reason for which the version exists.
I can't really see how you would make this work without _handles_, as in without _dynamically allocated entities_. On the other hand, dynamically allocate entities would be a huge waste of cpu cycles in the vast majority of cases. Therefore, I'm a bit puzzled.

> this could be adapted to an implementation that stores the underlying data in a compact representation, using similar techniques to what is done for the registry vector
> [...]
> Any thoughts on this approach and how feasible it would be to include in entt?

As it is (with shared pointers and whatnot), it looks technically unfeasible and a huge step back.+
However, I might change my mind if you explained in more details how one can port this implementation to something similar to what we have already.
I really can't easily fill the gap on my own, sorry.

> I would be interested in hearing about other suggestions for improving robustness as well.

Entity identifiers are fully customizable already. Moreover, I'm working to make the entity storage a proper storage which also is all in all similar to that of a random component `T`. I think you can easily tackle the problem in one way (custom type) or the other (custom mixin) without really changing the registry.
Have you already looked into these options?

**Reply by emsai** [reply]
Thanks for your detailed response.

> What do you mean exactly? Currently, it just wraps to 0. It's pretty well defined.

I guess I meant more that the program using entt will have undefined behaviour in this case, assuming it does not monitor the entity allocation and handle this case explicitly.

> I can't really see how you would make this work without handles, as in without dynamically allocated entities.

I think we are on the same page about this. Yes, the underlying entities need to be allocated dynamically in this case, as in the shared pointer example. More generally, I understand your hesitation to introduce a significant performance hit, and I don't think this feature needs to be used in applications that don't require it, but there are applications where the version and entity id overflow is a deal-breaker, since they require robust handling for these cases. What is the philosophy of entt in relation to these applications? Is this not considered a valid use case?

> However, I might change my mind if you explained in more details how one can port this implementation to something similar to what we have already.

What I meant by this is that the underlying entity data could be allocated dynamically with a dedicated pool of contiguous memory that could grow, managed by entt, if that was required as an optimization. I think this could be managed similarly to how the vector of entities is managed in the registry.

> Entity identifiers are fully customizable already. Moreover, I'm working to make the entity storage a proper storage which also is all in all similar to that of a random component T. I think you can easily tackle the problem in one way (custom type) or the other (custom mixin) without really changing the registry.

Can you clarify what you mean by making the entity storage a proper storage?

I don't think the customizability of entt is flexible enough to handle dynamically allocated entities, with pointers to these entities instead of the entities themselves in the registry vector (basically an extra layer of indirection). From what I can tell, all of the entity management in the registry is done using the predefined properties of the underlying uint32_t or uint64_t types. Do you agree, or am I missing something about how this customization can be done?

**Reply by skypjack** [reply]
> the program using entt will have undefined behaviour in this case, assuming it does not monitor the entity allocation and handle this case explicitly

I still don't understand how it's possible to be honest. Or better, since you've the version available, the only case is when you never cleanup pending identifiers that are stored aside for some reasons. Never or, well, with a very low frequency probably. What's your use case otherwise?

> the underlying entities need to be allocated dynamically in this case, as in the shared pointer example

An entity, an allocation. This is going to hurt the performance and to spawn tons of micro-allocations if you've many entities.

> What is the philosophy of entt in relation to these applications? Is this not considered a valid use case?

Well, I'm working to make the entity storage fully customizable for a reason. 🤷‍♂️ 

> Can you clarify what you mean by making the entity storage a proper storage?

The goal is to get rid of the vector in the registry and to replace it with a storage class (as in `entt::storage`).
Since you can control (and eventually suppress) deletion and creation of elements, it should give you the flexibility that you're looking for.
Moreover, storage classes support mixins and the like, that is, they are both customizable and easy to extend if you like.

That being said, can I ask you more details about your specific use case?
I can't easily wrap my mind around a problem that I don't know about and therefore it's hard to suggest the right approach (if any).

**Reply by Innokentiy-Alaytsev** [reply]
> An entity, an allocation. This is going to hurt the performance and to spawn tons of micro-allocations if you've many entities.

I _guess_ this can be resolved to some extent with some sort of pool allocator.

**Reply by emsai** [reply]
> Never or, well, with a very low frequency probably. What's your use case otherwise?

Correct, this affects any case that you hold entities over a long period of time relative to the rate at which you are deleting and creating entities. An example is storing entities in a map as described here: https://github.com/skypjack/entt/issues/337. Since previously deleted entities are prioritized for allocating new entities, with the default 32 bit ids, that only requires 4096 creations and deletions. Once this happens, if the version associated with the active entity happens to match the version of the stored entity, the code that uses the entity will be running in an invalid state, and there is no way to check for that.

Is the way around this just to centralize all deletions and include updates to additional data structures that contain entities there? Or to set up the entire program in a way that there is a bound on the number of creations and deletions that can happen across entity valid checks? This seems like it goes against the fact that anything should be able to hold entity values and use that as a key to the registry, even for long running tasks, and anything should be able to create entities. It also doesn't get to the root of the problem.

> The goal is to get rid of the vector in the registry and to replace it with a storage class (as in entt::storage).
Since you can control (and eventually suppress) deletion and creation of elements, it should give you the flexibility that you're looking for.

Very interesting, I am happy to hear this. It is not clear to me whether this will be enough to support reference counting and freeing up entities once reference counts have gone to zero, but I would be interested to try this out. Do you have an idea of when this change will be available?

> I guess this can be resolved to some extent with some sort of pool allocator.

Yes, that is what I was getting at, but that is more of an implementation detail. If we are focusing on the case that the user requires that the versions will not overflow, then sacrificing that for performance is not an option anyways.

**Reply by codylico** [reply]
> Since previously deleted entities are prioritized for allocating new entities, with the default 32 bit ids, that only requires 4096 creations and deletions.

...which seems to assume that the (12-bit _?_) version number is shared across all entity identifiers.

@skypjack Do you know if the version number is tracked per entity number (as opposed to a one version for all entities)? To me it seems that per-entity versions should prevent this sort of identifier collision, unless I'm missing something.

**Reply by emsai** [reply]
> ...which seems to assume that the (12-bit ?) version number is shared across all entity identifiers.

This is not the case. Here is an example of what actually happens:

Say that you repeatedly create a new entity and delete it, and don't do anything else. In that case, the created entity is added to the free list on deletion and reused with its version number incremented during the next creation. The version number is specific to the entity, and no other entity id values are involved. It take 4096 creations and deletions for the version number associated with this specific entity id to wrap. If you were to do this with two entities (two creations and deletions per step), then two entities would have their version number increased on each iteration.

**Reply by codylico** [reply]
so @emsai, to clarify, the issue is for external data structures that fail to keep up with the rapid deletion and creation?

**Reply by emsai** [reply]
Yes, rapid relative to the length of time that the data structures store the entity. For structures used in long running tasks, this does not need to be very fast.

**Reply by skypjack** [reply]
> Is the way around this just to centralize all deletions and include updates to additional data structures that contain entities there? Or to set up the entire program in a way that there is a bound on the number of creations and deletions that can happen across entity valid checks?

The way around this is very application specific. It doesn't depend on EnTT actually.
You can for example use a 64b identifier and have way more time before an overflow or you can set up a system that cleans up invalid entites from your map from time to time.

> Do you have an idea of when this change will be available?

Well, _soon-ish™️_ 🤷‍♂️ I'm actively working on it in my free time but it has lower priority than paid feature and things that I need at work.
In any case, it should be part of the next (non-minor) release. Dunno if you wanted to try `master` or just wait for a stable release.

**Reply by emsai** [reply]
Ok, thanks for the clarification, that makes sense. Good to know it is already in master, thanks for the heads up.
